File: selenium-project/src/main/java/com/master/seleniumproject/steps/login/LoginPageSteps.java
package com.master.seleniumproject.steps.login;

import com.master.seleniumproject.models.UserContext;
import com.master.seleniumproject.pages.login.LoginPage;
import org.openqa.selenium.WebDriver;

public class LoginPageSteps {

    private final LoginPage loginPage;

    public LoginPageSteps(WebDriver driver) {
        this.loginPage = new LoginPage(driver);
    }

    public void logIn(UserContext userContext) {
        this.fillUsernameInput(userContext.getUsername());
        this.fillPasswordInput(userContext.getPassword());
        this.clickLoginButton();
    }

    public void fillUsernameInput(String username) {
        this.loginPage.fillUsernameInput(username);
    }

    public void fillPasswordInput(String username) {
        this.loginPage.fillPasswordInput(username);
    }

    public void clickLoginButton() {
        this.loginPage.clickLoginButton();
    }
}

File: playwright-project/src/steps/ui/login/loginPage.steps.ts
import { Page, test } from '@playwright/test';
import { LoginPage } from '@pages/login/loginPage';
import { UserContext } from '@config/userContext';
import { BrowserSteps } from '@steps/ui/browser.steps';
import { BooksPage } from '@pages/books/booksPage';

export class LoginPageSteps extends BrowserSteps {
    constructor(
        page: Page,
        private loginPage: LoginPage = new LoginPage(page),
        private booksPage: BooksPage = new BooksPage(page),
    ) {
        super(page);
    }

    logIn = async (userContext: UserContext): Promise<void> => {
        await test.step(`Log in as '${userContext.email}' user`, async () => {
            await this.loginPage.openLoginPage();
            await this.fillUsernameField(userContext.email);
            await this.fillPasswordField(userContext.password);
            await this.clickLoginButton();

            await this.booksPage.waitForLogOutLinkIsVisible();
        });
    };

    verifyLoginPageIsOpen = async (): Promise<void> => {
        await test.step(`Verify 'Login' page is open`, async () => {
            await this.loginPage.waitForLoginPageOpened();
        });
    };

    /**
     * Fills 'Username/Email' input field
     * @param username
     */
    fillUsernameField = async (username: string): Promise<void> => {
        await test.step(`Fill 'Username' input by '${username}' on 'Login' page`, async () => {
            await this.loginPage.fillUsernameInput(username);
        });
    };

    /**
     * Fills 'Password' input field
     * @param password
     */
    fillPasswordField = async (password: string): Promise<void> => {
        await test.step(`Fill 'Password' input by '${password}' on 'Login' page`, async () => {
            await this.loginPage.fillPasswordInput(password);
        });
    };

    /**
     * Clicks 'Login' button
     */
    clickLoginButton = async (): Promise<void> => {
        await test.step(`Click 'Login' button on 'Login' page`, async () => {
            await this.loginPage.clickLoginButton();
        });
    };
}
